File: backend/app/api/github_connection.py
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Request, status

from app.models.github_connection import (
    GitHubConnectionConfig,
    GitHubConnectionPayload,
    GitHubConnectionTestResult,
)
from app.services.github_connection_service import GitHubConnectionService

router = APIRouter(prefix="/api/settings/github", tags=["github-settings"])


def get_github_connection_service(request: Request) -> GitHubConnectionService:
    return request.app.state.github_connection_service


@router.get("", response_model=GitHubConnectionConfig)
async def get_github_connection(
    service: Annotated[
        GitHubConnectionService, Depends(get_github_connection_service)
    ],
) -> GitHubConnectionConfig:
    return await service.get_connection()


@router.put("", response_model=GitHubConnectionConfig)
async def save_github_connection(
    payload: GitHubConnectionPayload,
    service: Annotated[
        GitHubConnectionService, Depends(get_github_connection_service)
    ],
) -> GitHubConnectionConfig:
    try:
        return await service.save_connection(payload)
    except ValueError as error:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(error),
        ) from error


@router.post("/test", response_model=GitHubConnectionTestResult)
async def test_github_connection(
    payload: GitHubConnectionPayload,
    service: Annotated[
        GitHubConnectionService, Depends(get_github_connection_service)
    ],
) -> GitHubConnectionTestResult:
    try:
        return await service.test_connection(payload)
    except ValueError as error:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(error),
        ) from error

File: frontend/src/api/githubConnection.ts
import { apiRequest } from "./client";
import type {
  GitHubConnectionConfig,
  GitHubConnectionPayload,
  GitHubConnectionTestResult,
} from "../types/githubConnection";

export const githubConnectionApi = {
  get(): Promise<GitHubConnectionConfig> {
    return apiRequest<GitHubConnectionConfig>("/api/settings/github");
  },

  save(payload: GitHubConnectionPayload): Promise<GitHubConnectionConfig> {
    return apiRequest<GitHubConnectionConfig>("/api/settings/github", {
      method: "PUT",
      body: payload,
    });
  },

  test(payload: GitHubConnectionPayload): Promise<GitHubConnectionTestResult> {
    return apiRequest<GitHubConnectionTestResult>("/api/settings/github/test", {
      method: "POST",
      body: payload,
    });
  },
};
